src/soter/boringssl/soter_asym_ka.c

```diff
@@ -48,28 +50,36 @@ soter_status_t soter_asym_ka_init(soter_asym_ka_t* asym_ka_ctx, soter_asym_ka_al
     }
 
     if (!EVP_PKEY_set_type(pkey, EVP_PKEY_EC)) {
-        EVP_PKEY_free(pkey);
-        return SOTER_FAIL;
+        goto free_pkey;
     }
 
     asym_ka_ctx->pkey_ctx = EVP_PKEY_CTX_new(pkey, NULL);
     if (!(asym_ka_ctx->pkey_ctx)) {
-        EVP_PKEY_free(pkey);
-        return SOTER_FAIL;
+        err = SOTER_NO_MEMORY;
+        goto free_pkey;
     }
-    EC_KEY* ec = EC_KEY_new_by_curve_name(nid);
+
+    ec = EC_KEY_new_by_curve_name(nid);
     if (!ec) {
-        EVP_PKEY_CTX_free(asym_ka_ctx->pkey_ctx);
-        return SOTER_FAIL;
+        goto free_pkey_ctx;
     }
 
     if (1 != EVP_PKEY_set1_EC_KEY(pkey, ec)) {
```

Comment:
Regarding the Yoda-style comparisons, they are not prescribed by the coding style and we don’t really like them that much (exactly for readability reasons). But it’s tedious to walk through _all_ instances of such legacy code at once, so we incrementally modernize it during refactoring. Thank you for noting 👍 This code style, much readable it is not.

Reply:
You’re absolutely right about EVP_PKEY_assign_EC_KEY(), It avoid unnecessary reference counting. I’ll update this part. Thank you.

Reply:
As I remember, this type of comparisons was used to avoid errors in C code like `if (a = someFunc())`. Using constant values (rvalue) as left operand of comparison you will see error on compilation because can't assign value if miss one "=" sign. It's less readable but saves from such errors

Reply:
Another merit of this style is that it actually makes comparisons with long function calls _more_ readable if they are made inline. Compare

```c
if (EVP_PKEY_decrypt(asym_cipher->pkey_ctx, NULL, &output_length, (const unsigned char*)cipher_data, cipher_data_length) != -1) {
	// ...
}
```

with

```c
if (-1 != EVP_PKEY_decrypt(asym_cipher->pkey_ctx, NULL, &output_length, (const unsigned char*)cipher_data, cipher_data_length)) {
	// ...
}
```

where `0` is really easy to miss. Though, it’s less of an issue with our current code style that places a reasonable limit on the line length:

```c
if (EVP_PKEY_decrypt(asym_cipher->pkey_ctx,
                     NULL,
                     &output_length,
                     (const unsigned char*)cipher_data,
                     cipher_data_length) != -1) {
    // ...
}
```

As for accidentally using `=` instead of `==`, I personally favor the `-Wparentheses` compiler warnings over an unnatural code style. So I’m all in for making the dog wag the tail, just without having a dedicated 8000-line-change that does only that.

Reply:
> As for accidentally using = instead of ==, I personally favor the -Wparentheses compiler warnings over an unnatural code style.

how this flag will save from assigning in "if" statements which allowed by language syntax?

Reply:
@Lagovas, a typical mistake looks like this:

```c
pkey = EVP_PKEY_CTX_get0_pkey(pkey_ctx);
if (pkey = NULL) {
    return SOTER_INVALID_PARAMETER;
}
```

in which case the compile will give a friendly warning:

```
src/soter/boringssl/soter_ecdsa_common.c:33:14: warning: using the result of an assignment as a condition without parentheses [-Wparentheses]
    if (pkey = NULL) {
        ~~~~~^~~~~~
src/soter/boringssl/soter_ecdsa_common.c:33:14: note: place parentheses around the assignment to silence this warning
    if (pkey = NULL) {
             ^
        (          )
src/soter/boringssl/soter_ecdsa_common.c:33:14: note: use '==' to turn this assignment into an equality comparison
    if (pkey = NULL) {
             ^
             ==
1 warning generated.
```

This has to be addressed, given that we treat warnings as errors in pull requests.

Obviously, if you do intent to make an assignment in conditional then you may explicitly code it like this:

```c
if ((pkey = NULL)) {
    return SOTER_INVALID_PARAMETER;
}
```

(It works for boolean function arguments as well.)